sycl/test/basic_tests/is_device_copyable.cpp

```diff
@@ -0,0 +1,70 @@
+// RUN: not %clangxx -fsycl -fsycl-device-only -fsyntax-only -Xclang -verify \
+// RUN: %s -I %sycl_include 2>&1 | FileCheck -check-prefix=CHECK-ERR %s
+
+// This test checks if compiler reports compilation error on an attempt to pass
+// a struct with type that is not device copyable as SYCL kernel parameter.
+// It also checks that the error is not emitted if the type trait
+// is_device_copyable is specialized for such types as true_type.
+
+#include <CL/sycl.hpp>
+
+using namespace cl::sycl;
+
+struct A { int i; };
+
+struct B {
+  int i;
+  B (int _i) : i(_i) {}
+  B (const B& x) : i(x.i) {}
+};
+
+// Identical to B, but will be declared as device copyable.
+struct BCopyable {
+  int i;
+  BCopyable (int _i) : i(_i) {}
+  BCopyable (const BCopyable& x) : i(x.i) {}
+};
+
+struct C : A {
+  const A C2;
+  C() : A{0}, C2{2}{}
+};
+
+struct D {
+  int i;
+  ~D();
+};
+
+// Identical to D, but will be declared as device copyable.
+struct DCopyable {
+  int i;
+  ~DCopyable();
+};
+
+template<> struct is_device_copyable<BCopyable> : std::true_type {};
+template<> struct is_device_copyable<DCopyable> : std::true_type {};
+
+void test() {
+  A IamGood;
+  IamGood.i = 0;
+  B IamBad(1);
+  BCopyable IamBadButCopyable(1);
+  C IamAlsoGood;
+  D IamAlsoBad{0};
+  DCopyable IamAlsoBadButCopyable{0};
+  queue Q;
+  Q.single_task<class kernel_capture_refs>([=] {
+    int A = IamGood.i;
+    int B = IamBad.i;
+
+    int C = IamBadButCopyable.i;
+    int D = IamAlsoBadButCopyable.i;
+
+    int E = IamAlsoGood.i;
+    int F = IamAlsoBad.i;
+  });
+}
+// CHECK-ERR: no expected directives found
+// CHECK-ERR-NEXT: diagnostics seen but not expected
+// CHECK-ERR-NEXT: is_device_copyable<B, void>
+// CHECK-ERR-NEXT: is_device_copyable<D, void>
```

Comment:
The diagnostics generated are telling you something important -- you're not testing what you think you're testing and this output is bad.

`no expected directives found` is coming from passing `-verify` on the RUN line but not telling the diagnostic verifier what diagnostics you expected to find. `diagnostics seen but not expected` is the diagnostic verifier telling you that there were actually diagnostics found.

I *think* you can remove the use of `FileCheck` and instead put the correct `// expected-error {{}}` markings in the file and let the diagnostic verifier handle checking the test results (preferred approach). Alternatively, you can remove the `-verify` and add the correct `CHECK` lines (only do this if you can't get `-verify` to work).

Reply:
@AaronBallman, haven't we agreed that "// expected-error {{}}" is not usable for this test, and the current approach is fine: https://github.com/intel/llvm/pull/3905#discussion_r649383664
This approach is also not invented by me, I've seen it used somewhere in clang or llvm.

> Alternatively, you can remove the -verify and add the correct CHECK lines (only do this if you can't get -verify to work

I don't understand the alternative idea. IMO, the current approach is reasonable and does not miss any cases where a) diagnistic is expected but not printed, b) when diagnostic is not expected but printed.

Reply:
Good point, we have (thank you for the reminder), but that doesn't mean the test isn't incorrect. Some of those `CHECK` lines are reporting errors due to the incorrect use of `-verify`, and that's what I'd like to see fixed. e.g., https://godbolt.org/z/6Eb7d66j5

The alternative idea is: remove `-verify` from the RUN line and remove the two CHECK lines that will stop showing up.